### PR TITLE
Converting create_instance to 2 separate functions

### LINKED ISSUE #5523: Simplify create_instance calls
Calling https://github.com/scrapy/scrapy/blob/bbfa185664cef79299b48cb0ae22065439bb07fc/scrapy/utils/misc.py#L144 with a crawler is needlessly verbose:

```python
create_instance(objclass, settings=None, crawler=crawler)
create_instance(objclass, settings=crawler.settings, crawler=crawler)  # functionally identical
```

It would be great if we could support a shorter alternative:

```python
create_instance(objclass, crawler=crawler)
```

It is not trivial to do in a backward-compatible way, given the function signature.

It may be also a good idea to log deprecation warnings when `settings` and `crawler` are used as positional arguments, as well as when `objcls` is used as a named parameter, so that we can eventually stop supporting that and move to a better signature:

```python
def create_instance(objcls, /, *args, crawler=None, settings=None, **kwargs): 
```

Although, looking at it now, I realize that such a signature would cause `settings` and `crawler` to be after `*args`, which could be confusing. Maybe it would be best to go for separate functions:

```python
def build_from_crawler(objcls, crawler, /, *args, **kwargs):
def build_from_settings(objcls, settings, /, *args, **kwargs):
```

### PR DESCRIPTION
**Pull request by sa2415**
Converted `create_instance` to 2 separate functions (`build_from_crawler` and `build_from_settings`).
Changed other calls to `create_instance` in other files.
Added tests for the new functions.

Fixes #5523, closes #5884, closes #6162.

### CODE REVIEW

**Comment by wRAR** [comment]
You should use pre-commit and you should run the tests locally before submitting.

**Comment by sa2415** [comment]
Hi - we attempted to run the tests locally but are running into issues. Will attempt to debug and will make another commit once resolved.

**Comment by monicaq21** [comment]
we made changes to the code that now passes all tests. Could this PR be marked as ready for review? Thank you! @wRAR

File: scrapy/addons.py
Comment (wRAR) [inline scrapy/addons.py:35]: Please remove the comments such as this one.

File: scrapy/addons.py
Comment (wRAR) [inline scrapy/addons.py:36]: It can't be None.

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:169]: Both new methods should get a proper complete docstring.

Comment (monicaq21) [reply scrapy/utils/misc.py]: Thank you for the feedback, the issues should be fixed now. I didn't have additional documentations to add for the methods, but I reformatted them similarly to other docstrings.

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:163]:
```suggestion
def build_from_crawler(objcls, crawler, /, *args, **kwargs):
```

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:187]:
```suggestion
def build_from_settings(objcls, settings, /, *args, **kwargs):
```

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:165]:
```suggestion
    """Construct a class instance using its ``from_crawler`` constructor.
```

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:167]:
```suggestion
    ``*args`` and ``**kwargs`` are forwarded to the constructor.
```

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:169]: Does it?

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:170]:
```suggestion
    Raises ``TypeError`` if the resulting instance is ``None``.
```

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:188]: All feedback for the `build_from_crawler` docstring also applies here.

File: scrapy/core/downloader/contextfactory.py
Comment (wRAR) [inline scrapy/core/downloader/contextfactory.py:175]: I don't think this comment should be removed.

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:188]: This is still not updated.

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:170]: As this is only used in the line 175, I'd just remove this var and used `crawler.settings` directly on that line.

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:171]: I think we shouldn't check `if crawler` and assume it's not None if this method was called.

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:192]: Similarly, I think we shouldn't check `if settings`.

File: scrapy/utils/misc.py
Comment (wRAR) [inline scrapy/utils/misc.py:185]:
```suggestion
    """Construct a class instance using its ``from_settings`` constructor.
```

**Comment by sa2415** [comment]
Hi @wRAR - thank you for the detailed feedback - is our work merge ready or would you like us to make more changes?

**Comment by monicaq21** [comment]
Hi! Sorry to bother you again, this PR is actually part of our final assignment that has significant credit, and having it merged would help us immensely. Could you check over the changes again and, if possible, merge it? Thank you so much @wRAR

File: tests/test_utils_misc/__init__.py
Comment (wRAR) [inline tests/test_utils_misc/__init__.py:109]: So this test only checks `from_crawler` while `test_build_from_settings` checks `from_crawler` and `from_settings`, I think this is an error, this one should check both while `test_build_from_settings` should only check `from_settings`.

File: tests/test_utils_misc/__init__.py
Comment (wRAR) [inline tests/test_utils_misc/__init__.py:114]: So this logic should be kept in `_test_with_crawler`.

File: tests/test_utils_misc/__init__.py
Comment (wRAR) [inline tests/test_utils_misc/__init__.py:131]: Including the fourth scenario.

File: tests/test_utils_misc/__init__.py
Comment (wRAR) [inline tests/test_utils_misc/__init__.py:98]: I also think the create_instance test should be kept as long as we keep the function itself.

File: scrapy/middleware.py
Comment (Gallaecio) [inline scrapy/middleware.py:70]: Can `crawler` be `None` at this point?

**Review by wRAR — APPROVED** [review]
Thanks!

**Review by monicaq21 — APPROVED** [review]

**Comment by sa2415** [comment]
Hi @wRAR - apologies for another ping - Was wondering if our PR can be merged with the base branch now if it's ready? Sorry for the slight rush, our deadline is in a day and we would receive credit if it is successfully merged...however, if there are still changes to be made, we will work accordingly to get those done as well. Thank you for your understanding!

**Comment by wRAR** [comment]
@sa2415 our policy is to merge contributor PRs after two approvals from maintainers so we can merge it once @Gallaecio approves it.

**Review by Gallaecio — APPROVED** [review]

**Comment by Gallaecio** [comment]
Awesome!

**Pull request merged**
